### PR TITLE
Add the ability to output vmdk via qemu-img

### PR DESCRIPTION
**Pull request by jpculp**
**Issue number:**
N/A

**Description of changes:**
This change allows users to edit the metadata in a variant's cargo.toml to output an OVA compatible vmdk through buildsys via qemu-img.

Double quotes were also added to several places in rpm2img to prevent unwanted globbing and word splitting (SC2086).

**Testing done:**

- Built aws-k8s-1.18 variant with no toml changes and received the standard img.lz4 output.

- Built aws-k8s-1.18 variant with ```image-format = "vmdk"``` added to the toml and received stream-optimized vmdk output for bottlerocket-aws-k8s-1.18-aarch64 and bottlerocket-aws-k8s-1.18-aarch64-data.

**Terms of contribution:**

By submitting this pull request, I agree that this contribution is dual-licensed under the terms of both the Apache License, version 2.0, and the MIT license.

### CODE REVIEW

File: Makefile.toml [hunk 1]
```diff
@@ -381,18 +381,11 @@ dependencies = ["build-variant"]
...
-  ${BUILDSYS_OUTPUT_DIR}/latest/${BUILDSYS_NAME_VARIANT}-data.img.lz4
-ln -snf ../${BUILDSYS_NAME_FULL}-boot.ext4.lz4 \
-  ${BUILDSYS_OUTPUT_DIR}/latest/${BUILDSYS_NAME_VARIANT}-boot.ext4.lz4
-ln -snf ../${BUILDSYS_NAME_FULL}-root.ext4.lz4 \
-  ${BUILDSYS_OUTPUT_DIR}/latest/${BUILDSYS_NAME_VARIANT}-root.ext4.lz4
-ln -snf ../${BUILDSYS_NAME_FULL}-root.verity.lz4 \
-  ${BUILDSYS_OUTPUT_DIR}/latest/${BUILDSYS_NAME_VARIANT}-root.verity.lz4
-ln -snf ../${BUILDSYS_NAME_FULL}-migrations.tar \
-  ${BUILDSYS_OUTPUT_DIR}/latest/${BUILDSYS_NAME_VARIANT}-migrations.tar
+for artifact in ${BUILDSYS_OUTPUT_DIR}/${BUILDSYS_NAME_FULL}*; do
```
Comment (jpculp) [inline Makefile.toml:386]: One pitfall of using a loop here is if a user builds a variant as an image, and then later rebuilds that same variant as a vmdk, if the build-id does not change it will link the old artifacts img.lz4 as well.

File: tools/buildsys/src/builder.rs [hunk 2]
```diff
@@ -89,6 +95,19 @@ impl VariantBuilder {
         let version_image = getenv("BUILDSYS_VERSION_IMAGE")?;
         let version_build = getenv("BUILDSYS_VERSION_BUILD")?;
         let output = getenv("BUILDSYS_OUTPUT_DIR")?;
+        let image_format = match image_format {
+            Some(ImageFormat::IMG) | None => String::from("img"),
+            Some(ImageFormat::VMDK) => String::from("vmdk"),
+        };
+        let image_options = match image_options {
+            Some(options_string) => options_string.join(","),
```
Comment (jahkeup) [inline tools/buildsys/src/builder.rs]: nit: `options_string` isn't a string here, let's avoid suggesting it is:
```suggestion
            Some(opts) => opts.join(","),
```

File: tools/buildsys/src/manifest.rs [hunk 3]
```diff
@@ -152,6 +162,15 @@ pub(crate) struct BuildPackage {
...
 pub(crate) struct BuildVariant {
     pub(crate) included_packages: Option<Vec<String>>,
+    pub(crate) image_format: Option<ImageFormat>,
+    pub(crate) image_options: Option<Vec<String>>,
+}
+
+#[derive(Deserialize, Debug)]
+#[serde(rename_all = "lowercase")]
+pub(crate) enum ImageFormat {
+    IMG,
```
Comment (jahkeup) [inline tools/buildsys/src/manifest.rs]: Can we refer to `IMG` instead as `Raw`? The sort of files that `IMG` is naming are typically called `raw` by tools and the various virtualization communities.

**Review by jahkeup — CHANGES_REQUESTED** [review]

File: Makefile.toml (hunk 1 above)
Comment (jahkeup) [reply Makefile.toml:386]: That might be desirable in some situations, for example: to build both a raw disk image *and* a VMDK.

File: tools/rpm2img [hunk 4]
```diff
@@ -90,43 +95,43 @@ rm -rf "${ROOT_MOUNT}"/var/lib "${ROOT_MOUNT}"/usr/share/licenses/*
...
 
 # Set the Bottlerocket variant, version, and build-id
-echo "PRETTY_NAME=\"Bottlerocket OS ${VERSION_ID}\"" >> ${ROOT_MOUNT}/${SYS_ROOT}/usr/lib/os-release
-echo "VARIANT_ID=${VARIANT}" >> ${ROOT_MOUNT}/${SYS_ROOT}/usr/lib/os-release
-echo "VERSION_ID=${VERSION_ID}" >> ${ROOT_MOUNT}/${SYS_ROOT}/usr/lib/os-release
-echo "BUILD_ID=${BUILD_ID}" >> ${ROOT_MOUNT}/${SYS_ROOT}/usr/lib/os-release
+echo "PRETTY_NAME=\"Bottlerocket OS ${VERSION_ID}\"" >> "${ROOT_MOUNT}/${SYS_ROOT}/usr/lib/os-release"
+echo "VARIANT_ID=${VARIANT}" >> "${ROOT_MOUNT}/${SYS_ROOT}/usr/lib/os-release"
+echo "VERSION_ID=${VERSION_ID}" >> "${ROOT_MOUNT}/${SYS_ROOT}/usr/lib/os-release"
+echo "BUILD_ID=${BUILD_ID}" >> "${ROOT_MOUNT}/${SYS_ROOT}/usr/lib/os-release"
```
Comment (bcressey) [inline tools/rpm2img:142]: I'd like to have all these unrelated shellcheck fixes in a separate commit.

File: tools/buildsys/src/builder.rs (hunk 2 above)
Comment (bcressey) [inline tools/buildsys/src/builder.rs]: `image_opts` feels like the wrong abstraction boundary here - it's not consistent with the other handling of low-level details that takes place in `rpm2img`.

Let's hardcode the stream-optimized options in `rpm2img` when building VMDKs. We can name the format `stream-vmdk` if we need to distinguish it from ordinary VMDK for some reason, but I don't think we do.

File: Makefile.toml (hunk 1 above)
Comment (bcressey) [reply Makefile.toml:386]: I have a fix coming soon for #1027 that should address this; artifacts from a previous `buildsys` invocation will be cleaned up prior to the next, so we shouldn't hit the situation where we have both types at once.

**Review by bcressey — CHANGES_REQUESTED** [review]

File: Makefile.toml (hunk 1 above)
Comment (zmrow) [reply Makefile.toml:386]: @jpculp Let's remove this loop.  If another solution is coming we don't need to fix the "problem" here.

File: tools/rpm2img [hunk 5]
```diff
@@ -3,22 +3,26 @@
 set -eu -o pipefail
 shopt -qs failglob
 
+OUTPUT_FMT="raw"
+OUTPUT_OPTS="none"
```
Comment (zmrow) [inline tools/rpm2img]: Let's eliminate this option for now

File: tools/rpm2img [hunk 6]
```diff
@@ -206,16 +210,34 @@ dd if="${BOTTLEROCKET_DATA}" of="${DATA_IMAGE}" conv=notrunc bs=1M seek=1
...
+if [[ ${OUTPUT_FMT} != "raw" ]]; then
+  QEMU_PARAMS="-f raw -O ${OUTPUT_FMT}"
+
+  # Stream optimization is required for creating an Open Virtual Appliance (OVA).
+  if [[ ${OUTPUT_FMT} == "vmdk" ]]; then
+    OUTPUT_OPTS="subformat=streamOptimized"
+  fi
+
+  if [[ ${OUTPUT_OPTS} != "none" ]]; then
+    QEMU_PARAMS="${QEMU_PARAMS} -o ${OUTPUT_OPTS}"
```
Comment (zmrow) [inline tools/rpm2img]: I think this is a bit over complicated.  We can simplify it a bit by eliminating the multiple extra variables and having the qemu command hardcoded for now.  If you want to encapsulate all qemu args in a single variable that can be re-used, that's fine too.

IMHO it also might help to match on a positive case first to make it simpler to read:

```
if [[ ${OUTPUT_FMT} == "vmdk" ]]; then
    # Don't need to use this variable, could just hard code the commands
    QEMU_OPTS="convert -f raw -O vmdk -o subformat=streamOptimized"

    qemu-img "${QEMU_OPTS}" .....

# Can use elif's later for additional output formats
else
    ....
fi
```__

**Review by zmrow — CHANGES_REQUESTED** [review]

File: Makefile.toml (hunk 1 above)
Comment (jpculp) [reply Makefile.toml:386]: @bcressey  can correct me, but I think his solution is just for cleaning up old artifacts, such as in the situation I mentioned above. This loop would still be necessary for generating the proper symlinks under the build/images/latest directory.

File: tools/buildsys/src/builder.rs (hunk 2 above)
Comment (jpculp) [reply tools/buildsys/src/builder.rs]: Makes sense to me, but it looks like we're just going to remove image_options for now.

File: tools/rpm2img (hunk 6 above)
Comment (jpculp) [reply tools/rpm2img]: I've updated my commits with something in-between your suggestion and the original commit.

File: Makefile.toml (hunk 1 above)
Comment (zmrow) [reply Makefile.toml:386]: Ah yeah you're right!

**Review by zmrow — APPROVED** [review]
:hamburger:

File: tools/rpm2img [hunk 7]
```diff
@@ -3,22 +3,25 @@
...
-DISK_IMAGE_NAME="${FILENAME_PREFIX}.img.lz4"
-BOOT_IMAGE_NAME="${FILENAME_PREFIX}-boot.ext4.lz4"
-VERITY_IMAGE_NAME="${FILENAME_PREFIX}-root.verity.lz4"
-ROOT_IMAGE_NAME="${FILENAME_PREFIX}-root.ext4.lz4"
-DATA_IMAGE_NAME="${FILENAME_PREFIX}-data.img.lz4"
+DISK_IMAGE_BASENAME="${FILENAME_PREFIX}"
+BOOT_IMAGE_BASENAME="${FILENAME_PREFIX}-boot"
+VERITY_IMAGE_BASENAME="${FILENAME_PREFIX}-root"
+ROOT_IMAGE_BASENAME="${FILENAME_PREFIX}-root"
+DATA_IMAGE_BASENAME="${FILENAME_PREFIX}-data"
```
Comment (bcressey) [inline tools/rpm2img]: It's weird to me that `VERITY_IMAGE_BASENAME` and `ROOT_IMAGE_BASENAME` have the same value. Maybe incorporate part of the extension?

```suggestion
DISK_IMAGE_BASENAME="${FILENAME_PREFIX}"
DATA_IMAGE_BASENAME="${FILENAME_PREFIX}-data"

BOOT_IMAGE_BASENAME="${FILENAME_PREFIX}-boot.ext4"
ROOT_IMAGE_BASENAME="${FILENAME_PREFIX}-root.ext4"
VERITY_IMAGE_BASENAME="${FILENAME_PREFIX}-root.verity"
```

Then the final extension (`.img.lz4`, `.vmdk`, `.lz4`) would be provided on the same line where we invoke the tool that converts it to that format, which would keep the clarity improvement you're after.

Comment (jpculp) [reply tools/rpm2img]: I can see how that would be strange. I've gone ahead and amended the commit so that we only use BASENAME for the disk and data images.

File: tools/rpm2img [hunk 8]
```diff
@@ -206,14 +211,28 @@ dd if="${BOTTLEROCKET_DATA}" of="${DATA_IMAGE}" conv=notrunc bs=1M seek=1
...
+  qemu-img convert -f raw -O vmdk -o subformat=streamOptimized "${DISK_IMAGE}" "${OUTPUT_DIR}/${DISK_IMAGE_BASENAME}.vmdk"
+  qemu-img convert -f raw -O vmdk -o subformat=streamOptimized "${DATA_IMAGE}" "${OUTPUT_DIR}/${DATA_IMAGE_BASENAME}.vmdk"
+  chown 1000:1000 "${OUTPUT_DIR}/${DISK_IMAGE_BASENAME}.vmdk" \
+    "${OUTPUT_DIR}/${DATA_IMAGE_BASENAME}.vmdk"
+else
+  qemu-img convert -f raw -O "${OUTPUT_FMT}" "${DISK_IMAGE}" "${OUTPUT_DIR}/${DISK_IMAGE_BASENAME}.${OUTPUT_FMT}"
+  qemu-img convert -f raw -O "${OUTPUT_FMT}" "${DATA_IMAGE}" "${OUTPUT_DIR}/${DATA_IMAGE_BASENAME}.${OUTPUT_FMT}"
+  chown 1000:1000 "${OUTPUT_DIR}/${DISK_IMAGE_BASENAME}.${OUTPUT_FMT}" \
+    "${OUTPUT_DIR}/${DATA_IMAGE_BASENAME}.${OUTPUT_FMT}"
+fi
```
Comment (bcressey) [inline tools/rpm2img]: If we don't currently expect to take this path, I'd prefer to just error out:

```suggestion
else
  echo "unexpected image format '${OUTPUT_FMT}'" >&2
  exit 1
fi
```

**Review by bcressey — COMMENTED** [review]
LGTM, just one small concern.

File: tools/rpm2img (hunk 8 above)
Comment (jpculp) [reply tools/rpm2img]: After some further discussion, I removed the else entirely and added a case statement to check for valid output formats immediately after the argument parsing. To support additional formats in the future, they will have to be added to the case statement and an additional elif will have to be appended to this block.

**Review by jahkeup — APPROVED** [review]

**Review by zmrow — APPROVED** [review]
:medal_military:

**Review by bcressey — APPROVED** [review]
